clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1581,86 +1581,131 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
     const auto *PropDecl =
         cast<ClassTemplateSpecializationDecl>(PropTy->getAsRecordDecl());
-    if (PropDecl->getTemplateArgs().size() != 1) {
-      SemaRef.Diag(Loc, diag::err_sycl_invalid_property_list_param_number)
-          << "buffer_location";
-      return;
-    }
+    if (PropDecl->getTemplateArgs().size() != 1)
+      return SemaRef.Diag(Loc,
+                          diag::err_sycl_invalid_property_list_param_number)
+             << "buffer_location";
+
     const auto BufferLoc = PropDecl->getTemplateArgs()[0];
-    if (BufferLoc.getKind() != TemplateArgument::ArgKind::Integral) {
-      SemaRef.Diag(Loc,
-                   diag::err_sycl_invalid_accessor_property_list_template_param)
-          << /*buffer_location*/ 2 << /*non-negative integer*/ 2;
-      return;
-    }
+    if (BufferLoc.getKind() != TemplateArgument::ArgKind::Integral)
+      return SemaRef.Diag(
+                 Loc,
+                 diag::err_sycl_invalid_accessor_property_list_template_param)
+             << /*buffer_location*/ 2 << /*non-negative integer*/ 2;
+
     int LocationID = static_cast<int>(BufferLoc.getAsIntegral().getExtValue());
-    if (LocationID < 0) {
-      SemaRef.Diag(Loc,
-                   diag::err_sycl_invalid_accessor_property_list_template_param)
-          << /*buffer_location*/ 2 << /*non-negative integer*/ 2;
-      return;
-    }
+    if (LocationID < 0)
+      return SemaRef.Diag(
+                 Loc,
+                 diag::err_sycl_invalid_accessor_property_list_template_param)
+             << /*buffer_location*/ 2 << /*non-negative integer*/ 2;
+
+    return false;
   }
 
-  void checkAccessorType(QualType Ty, SourceRange Loc) {
+  bool checkAccessorType(QualType Ty, SourceRange Loc) {
     assert(Util::isSyclAccessorType(Ty) &&
            "Should only be called on SYCL accessor types.");
 
     const RecordDecl *RecD = Ty->getAsRecordDecl();
+    bool isInvalid = false;
     if (const ClassTemplateSpecializationDecl *CTSD =
             dyn_cast<ClassTemplateSpecializationDecl>(RecD)) {
       const TemplateArgumentList &TAL = CTSD->getTemplateArgs();
+      if (TAL.size() < 5)
+        return SemaRef.Diag(
+            Loc.getBegin(),
+            diag::err_sycl_invalid_accessor_template_param_number);
+
+      // A typename specifying the data type that the accessor
+      // is providing access to. Ensure SYCL type restrictions
+      // are applied.
       TemplateArgument TA = TAL.get(0);
-      const QualType TemplateArgTy = TA.getAsType();
+      llvm::DenseSet<QualType> Visited;
+      checkSYCLType(SemaRef, TA.getAsType(), Loc, Visited);
+
+      // An integer specifying the dimensionality of the accessor.
+      TA = TAL.get(1);
+      if (TA.getKind() != TemplateArgument::Integral)
+        isInvalid = SemaRef.Diag(Loc.getBegin(),
+                                 diag::err_sycl_invalid_accessor_template_param)
+                    << /*third*/ 0 << /*access::mode*/ 0;
```

Comment:
Comments look wrong here.

Reply:
Btw, the off-by-two nature of `0` vs `second` is a bit jarring every time I run into it. I don't know what to do about it, but thought it was a bit strange nonetheless.